### PR TITLE
feat: handle if user logged in to other instance and visited a link

### LINKED ISSUE #15114: Managing external links to internal parts of the app when we have app.posthog.com and eu.posthog.com
(I'm happy to build this but would be good to get others input on it)

TL;DR if you go to a url such as https://app.posthog.com/login?next=/dashboard_templates/user_research when you are only logged in on eu.posthog.com not app.posthog then show a message and autoredirect after a few seconds

## Is your feature request related to a problem?

Problem:
- We often want to link to areas of the app (e.g. the settings page or specific dashboard templates) from our own website.
- We also want others users to be easily able to share links to internal parts of our app (e.g. when supporting each other via the community, sharing links on twitter to new features)
- However, we have two subdomains - eu.posthog.com and app.posthog.com and currently the links don't work across them

## Describe the solution you'd like

- When a user tries to visit a link on one of the subdomains which isn't `/`, `/signup` or `/login`, AND they are logged in to the other subdomain, then shows a redirecting message and redirect them after a few seconds
  - Why this? It means us and customers can both share any link and it should just work regardless if it's eu or app.posthog.com
  - We could also add a query parameter which skips the redirect message e.g. if linking from posthog.com and we know we want to go to the one they are logged in on

- What if users have an app.posthog.com and an eu.posthog.com login?
  - By default, do nothing and just take them to the link they clicked on. As most of these links will probably be internal links shared to the correct domain. If they need to, they can change the url directly as most of our ICP customers are savvy enough to realize and do this.
  - In in the unlikely case there's lots of confusion around people being logged into both and they are clicking on links on our marketing site, we could either add a query parameter to explicitly show a redirect message OR have a dedicated subdomain e.g. redirect.posthog.com However, this would only work when people are willing to manually change the links (most likely when we add links to our posthog.com site)

## Describe alternatives you've considered

- the url already automatically translates https://eu.posthog.com/feature_flags/5678 to https://eu.posthog.com/login?next=/feature_flags/5678 if they are not logged in. And when they change the data region from EU to US it automatically takes them to the link. So we could just do nothing however I think it's a better experience
- redirect.posthog.com or posthog.com/redirect However, this would only work when people are willing to manually change the links

### PR DESCRIPTION
**Pull request by lharries**
## Problem

Closes https://github.com/PostHog/posthog/issues/15114

## Changes

If you try and go to an (app / eu).posthog.com link, and you're not logged in (detected by them being sent to /login) but you are logged in to the other instance then show a toast message and auto-redirect the user

It won't work on the sign up page (in case you are actually wanting to sign up for the other URL), but it works on all the other pages

I only update the subdomain in case somehow someone manages to inject a cookie and then redirects them to a different site

https://user-images.githubusercontent.com/22766134/233104494-ec6953b0-4d53-478c-9c04-47002dfe63a4.mp4

## How I've tested it

Added a cypress test for the toast message (currently turned off in this PR but I'll turn them on again when I remove the flag - haven't quickly worked out how to do e2e tests with mocked feature flags)

Instructions for testing locally with a local host

### Summary
🍪🚀🧪

This pull request implements a feature to redirect users to the correct subdomain if they are already logged in to a different instance of PostHog. It modifies the `Login` component, adds a new function `redirectIfLoggedInOtherInstance`, and creates a new test file `auth-redirect.cy.ts`.

> _Oh we are the PostHog crew and we code with skill and pride_
> _We test our features well and we check the cookies right_
> _We use `redirectIfLoggedInOtherInstance` to send the users on their way_
> _And we sing this shanty chorus as we work and as we play_

### Walkthrough
*  Add a new function `redirectIfLoggedInOtherInstance` to check and redirect the user based on their login status and cookie values ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-3441d02d27b9a9e28c963bf482fb13dc7ff2edd85e692f096391e4c801f30fa9R1-R95), [link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-902ae2d786797e2b8f1a2a4ba3c2c5fd0ff2c33398aefa35ed54ec9f0dec04e3R16), [link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-902ae2d786797e2b8f1a2a4ba3c2c5fd0ff2c33398aefa35ed54ec9f0dec04e3R58-R65))
  - The function reads the `ph_current_instance` and `is-logged-in` cookies and compares the current subdomain with the logged in subdomain ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-3441d02d27b9a9e28c963bf482fb13dc7ff2edd85e692f096391e4c801f30fa9R1-R95))
  - The function handles the cases where the cookies are missing or invalid and decides if a redirection is needed ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-3441d02d27b9a9e28c963bf482fb13dc7ff2edd85e692f096391e4c801f30fa9R1-R95))
  - The function uses the `lemonToast` component to show an info toast with a cancel button and a timeout before redirecting the user ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-3441d02d27b9a9e28c963bf482fb13dc7ff2edd85e692f096391e4c801f30fa9R1-R95))
  - The function updates the URL with the new subdomain and redirects the user when the toast is closed or timed out ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-3441d02d27b9a9e28c963bf482fb13dc7ff2edd85e692f096391e4c801f30fa9R1-R95))
  - The function has some comments explaining how to test it locally and the limitations of the current implementation ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-3441d02d27b9a9e28c963bf482fb13dc7ff2edd85e692f096391e4c801f30fa9R1-R95))
* Call the `redirectIfLoggedInOtherInstance` function inside the `Login` component's `useEffect` hook to try to redirect the user when the component mounts ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-902ae2d786797e2b8f1a2a4ba3c2c5fd0ff2c33398aefa35ed54ec9f0dec04e3R58-R65))
  - The hook also catches and logs any errors that might occur during the redirection process ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-902ae2d786797e2b8f1a2a4ba3c2c5fd0ff2c33398aefa35ed54ec9f0dec04e3R58-R65))
* Add a new test file `auth-redirect.cy.ts` to contain two test cases for checking the redirection function ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-da490edddf260a4a23e10fc16ccdfb3291819eca25aaa49f0578e74f76878d68R1-R42))
  - The test cases use Cypress commands to visit the login page, set the cookies, and reload the page ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-da490edddf260a4a23e10fc16ccdfb3291819eca25aaa49f0578e74f76878d68R1-R42))
  - The test cases assert that the expected info toast is shown with the subdomain name ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-da490edddf260a4a23e10fc16ccdfb3291819eca25aaa49f0578e74f76878d68R1-R42))
  - The test cases are skipped for now because the actual redirection is not working with Cypress ([link](https://github.com/PostHog/posthog/pull/15119/files?diff=unified&w=0#diff-da490edddf260a4a23e10fc16ccdfb3291819eca25aaa49f0578e74f76878d68R1-R42))

### CODE REVIEW

File: frontend/src/scenes/authentication/redirectToLoggedInInstance.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// cookie values
+const PH_CURRENT_INSTANCE = 'ph_current_instance'
+const IS_LOGGED_IN = 'is-logged-in'
+
+const REDIRECT_TIMEOUT = 2500
+
+const SUBDOMAIN_TO_NAME = {
+    eu: 'EU cloud',
+    app: 'US cloud',
+}
```
Comment (Twixes) [inline frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:36]:
```suggestion
} as const
```
for some slightly nicer typing

File: frontend/src/scenes/authentication/redirectToLoggedInInstance.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+export function redirectIfLoggedInOtherInstance(): (() => void) | undefined {
+    const currentSubdomain = window.location.hostname.split('.')[0]
+
+    const loggedInSubdomain = getCookie(PH_CURRENT_INSTANCE)
+        ?.replace('http://', '')
+        ?.replace('https://', '')
+        .split('.')[0]
```
Comment (Twixes) [inline frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:44]:
```suggestion
    const currentInstance = getCookie(PH_CURRENT_INSTANCE)
    const loggedInSubdomain = currentInstance ? new URL(currentInstance).host : null
```

File: frontend/src/scenes/authentication/redirectToLoggedInInstance.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+// cookie values
+const PH_CURRENT_INSTANCE = 'ph_current_instance'
+const IS_LOGGED_IN = 'is-logged-in'
+
+const REDIRECT_TIMEOUT = 2500
+
+const SUBDOMAIN_TO_NAME = {
+    eu: 'EU cloud',
+    app: 'US cloud',
```
Comment (Twixes) [inline frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:35]: They're all Cloud, so we can just have "EU" and "US" here, right?

File: frontend/src/scenes/authentication/redirectToLoggedInInstance.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        let cancelClicked = false
+
+        const closeToastAction = (): void => {
+            if (cancelClicked) {
+                return
+            }
+            window.location.assign(newUrl.href)
+        }
+
+        lemonToast.info('Redirecting to your logged in account on the ' + SUBDOMAIN_TO_NAME[loggedInSubdomain], {
```
Comment (Twixes) [inline frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:75]:
```suggestion
        lemonToast.info(`Redirecting to your logged-in account in the Cloud ${SUBDOMAIN_TO_NAME[loggedInSubdomain]} region`, {
```

File: frontend/src/scenes/authentication/redirectToLoggedInInstance.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    const loggedInSubdomain = getCookie(PH_CURRENT_INSTANCE)
+        ?.replace('http://', '')
+        ?.replace('https://', '')
+        .split('.')[0]
+
+    // when they are logged out, ph_instance is not removed.
+    // therefore, use is-logged-in cookie to determine if they are logged in
+    // note: this seems to be set when logged in and removed when not
+    // however I can't find the code where it is set and removed so potentially unreliable
```
Comment (Twixes) [inline frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:49]: I don't have the `is-logged-in` cookie set on either Cloud, and I can't find any code for it either, so I don't think we should use it

**Review by Twixes — CHANGES_REQUESTED** [review]
This makes a ton of sense, just I can't say if `is-logged-in` is an actual thing that works (+ a few minor stylistic comments)

File: frontend/src/scenes/authentication/Login.tsx [hunk 6]
```diff
@@ -49,10 +52,21 @@ export function Login(): JSX.Element {
...
     const passwordInputRef = useRef<HTMLInputElement>(null)
     const isPasswordHidden = precheckResponse.status === 'pending' || precheckResponse.sso_enforcement
 
+    useEffect(() => {
+        try {
+            if (featureFlags[FEATURE_FLAGS.AUTO_REDIRECT]) {
+                redirectIfLoggedInOtherInstance()
+            }
+        } catch (e) {
+            console.error('unable to redirect to logged in instance', e)
```
Comment (Twixes) [inline frontend/src/scenes/authentication/Login.tsx:66]: We should probably not just `console.error()` but send to Sentry (`captureException()`)

File: frontend/src/scenes/authentication/redirectToLoggedInInstance.ts (hunk 2 above)
Comment (lharries) [reply frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:44]: Actually this doesn't seem to work as the current instance isn't a full url

Comment (Twixes) [reply frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:44]: Oh right, I missed the subdomain split part. Weird that `new URL()` doesn't work though, all it needs is a protocol + host AFAIK, thought current instance does include both

Comment (lharries) [reply frontend/src/scenes/authentication/redirectToLoggedInInstance.ts:44]: Realized my example cookie was wrong, will try again

**Review by Twixes — APPROVED** [review]

**Pull request merged**